pkg/ruler/manager/memstore.go

```diff
@@ -0,0 +1,380 @@
+	if ruleKey == "" {
+		level.Error(m.logger).Log("msg", "MemStoreQuerier.Select called in an unexpected fashion without alertname or ALERTS_FOR_STATE labels")
+		return storage.NoopSeriesSet()
+	}
+
+	var rule *rules.AlertingRule
+
+	// go fetch the rule via the alertname
+	for _, x := range m.mgr.AlertingRules() {
+		if x.Name() == ruleKey {
+			rule = x
+			break
+		}
+	}
+
+	// should not happen
+	if rule == nil {
+		level.Error(m.logger).Log("msg", "failure trying to restore for state for untracked alerting rule", "name", ruleKey)
+		return storage.NoopSeriesSet()
+	}
+
+	level.Debug(m.logger).Log("msg", "restoring for state via evaluation", "rule", ruleKey, "tenant", m.MemStore.userID)
+
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	cache, ok := m.rules[ruleKey]
+
+	// no timestamp results are cached for this rule at all; Create it.
+	if !ok {
+		cache = NewRuleCache(m.metrics)
+		m.rules[ruleKey] = cache
+	}
+
+	smpl, cached := cache.Get(m.ts, ls)
+	if cached {
+		m.metrics.CacheHits.WithLabelValues(m.userID).Inc()
+		level.Debug(m.logger).Log("msg", "result cached", "rule", ruleKey, "tenant", m.MemStore.userID)
+		// Assuming the result is cached but the desired series is not in the result, it wouldn't be considered active.
+		if smpl == nil {
+			return storage.NoopSeriesSet()
+		}
+
+		// If the labelset is cached we can consider it active. Return the for state sample active immediately.
+		return series.NewConcreteSeriesSet(
+			[]storage.Series{
+				series.NewConcreteSeries(smpl.Metric, []model.SamplePair{
+					{Timestamp: model.Time(util.TimeToMillis(m.ts)), Value: model.SampleValue(smpl.V)},
+				}),
+			},
+		)
+	}
+
+	// see if alert condition had any inhabitants at ts-forDuration. We can assume it's still firing because
+	// that's the only condition under which this is queried (via RestoreForState).
+	vec, err := m.queryFunc(m.ctx, rule.Query().String(), m.ts.Add(-rule.HoldDuration()))
+	if err != nil {
+		level.Info(m.logger).Log("msg", "error querying for rule", "rule", ruleKey, "tenant", m.MemStore.userID, "err", err.Error())
+		m.metrics.Evaluations.WithLabelValues(statusFailure, m.userID).Inc()
+		return storage.NoopSeriesSet()
+	}
+	m.metrics.Evaluations.WithLabelValues(statusSuccess, m.userID).Inc()
+	level.Debug(m.logger).Log("msg", "rule state successfully restored", "rule", ruleKey, "tenant", m.MemStore.userID, "len", len(vec))
+
+	// translate the result into the ALERTS_FOR_STATE series for caching,
+	// considered active & written at the timetamp requested
+	forStateVec := make(promql.Vector, 0, len(vec))
+	for _, smpl := range vec {
+
+		ts := util.TimeToMillis(m.ts)
+
+		forStateVec = append(forStateVec, promql.Sample{
+			Metric: ForStateMetric(smpl.Metric, rule.Name()),
+			Point: promql.Point{
+				T: ts,
+				V: float64(ts),
+			},
+		})
+
+	}
+
+	// cache the result of the evaluation at this timestamp
+	cache.Set(m.ts, forStateVec)
+
+	// Finally return the series if it exists
+	smpl, ok = cache.Get(m.ts, ls)
```

Comment:
This seems redundant, didn't we just put this timestamp in the cache 2 lines ago?

Reply:
We previously stored the entire vector resulting from evaluating the rule. When we request it from the cache we're looking for one possible element from that vector. We could iterate through the vector and label-match without using the cache, but it'd result in a lot of duplicate logic that's better contained in the cache IMO.